### ISSUE TITLE
How to display Top2Vec Model in HDBSCAN or UMAP ?

### ISSUE THREAD

**Issue by i4meta** [ddangelov/Top2Vec#133]
Hello,

Forgive me for the newbie question, but having successfully built and saved a Top2Vec model:

         How can a saved Top2Vec model be viewed (visually rendered) in HDBSCAN or UMAP?

I may be over looking the obvious, but in reading through the documentation and Googling for answers nothing has jumped out so far.

Most grateful,

Chris

**Comment by MattiaPance** [comment]
+1

**Comment by nusipek** [comment]
+1

**Comment by hp0404** [comment]
I tried following the source code (see [here](https://github.com/ddangelov/Top2Vec/blob/95b19307dd6b1d50c092272d4eb046fc40ed6a78/top2vec/Top2Vec.py#L350-L355)) and defined `umap` instance myself, but it would help if `umap` could be accessed through the model, e.g. `self.umap` 

---

Example: 

```python
model = Top2Vec.load("./../models/model")
umap_args = {
    "n_neighbors": 15,
    "n_components": 2, # 5 -> 2 for plotting 
    "metric": "cosine",
}
umap_model = umap.UMAP(**umap_args).fit(model._get_document_vectors(norm=False))
umap.plot.points(umap_model, labels=model.doc_top)
```

**Comment by doodyology** [comment]
I believe the following code will work.

You can get the document vectors from the trained model object:
```python
vectors = model.document_vectors
```

And then fit your own 2D UMAP model:
```python
reduced2d = umap.UMAP(n_neighbors=15, n_components=2, metric='cosine', verbose=True).fit_transform(vectors)
```

And then you can plot the data in 2D space:
```python
x, y = reduced2d[:,0], reduced2d[:,1]
plt.scatter(x,y)
```

**Comment by i4meta** [comment]
@doodyology - Not following your last line . . . "plt.scatter(x,y)" as I get 'plt' is not defined

**Comment by doodyology** [comment]
> 
> 
> @doodyology - Not following your last line . . . "plt.scatter(x,y)" as I get 'plt' is not defined

That's the standard syntax for the plotting library matplotlib. So, you would have to import matlpotlib as such:

```python
import matplotlib.pyplot as plt
```

**Comment by WKSu** [comment]
Hi I added a way to save and retrieve these models when they are generated so you can load them later in https://github.com/ddangelov/Top2Vec/pull/149. I believe running these commands again after generating the model already might create different results due to the stochastic nature of these algorithms, so it might be nicer to retrieve the initial instance instead.

Visualization after loading them can then be performed as mentioned in their respective docs. Do want to say that the default umap arguments specify 5 components (dimensions) which might be difficult to visualize. You could run it again with only 2 dimensions to showcase the clusters but up to you.

**Comment by pheewma** [comment]
Combined @WKSu and @hp0404 's methods works well! However, I ended up reducing the number of topics from something like 3000+ topics to something more manageable, but when I plot on UMAP, the labels are for the original (3000+ topics) model. How would I go about plotting with the labels for the reduced topics?

**Comment by doodyology** [comment]
> 
> 
> Combined @WKSu and @hp0404 's methods works well! However, I ended up reducing the number of topics from something like 3000+ topics to something more manageable, but when I plot on UMAP, the labels are for the original (3000+ topics) model. How would I go about plotting with the labels for the reduced topics?

I believe you can get the reduced vectors with:

```python
reduced_vectors = model.topic_vectors_reduced
```

**Comment by pheewma** [comment]
> > Combined @WKSu and @hp0404 's methods works well! However, I ended up reducing the number of topics from something like 3000+ topics to something more manageable, but when I plot on UMAP, the labels are for the original (3000+ topics) model. How would I go about plotting with the labels for the reduced topics?
> 
> I believe you can get the reduced vectors with:
> 
> ```python
> reduced_vectors = model.topic_vectors_reduced
> ```

Thanks for getting me started! The solution I found was similar though not exactly what you stated. @hp0404 uses: 

`model.doc_top`

to retrieve the topic labels

If you want to use the reduced document labels the relevant method is:

`model.doc_top_reduced`


So end result, the code block looks something like:

```
import umap.plot
model = Top2Vec.load('model_name')
umap.plot.points(umap_model, labels=model.doc_top_reduced)
```

Thanks again for the help!
